### PR TITLE
Foundation overlay cleanup

### PR DESCRIPTION
**Pull request by phausler**
This is intended as a non functional change set. Primarily to keep maintainers sane and the layout organized into isolated areas of concern. The Foundation.swift file worked fine when there were only a handful of items to work with; but as Foundation's swift interoperability layer has grown it has been quite evident that we will run the risk of that file becoming a dumping ground for disparate ideas. The changes here are just copy/paste from the original Foundation.swift.gyb to a file for each idea group. In most cases this is a single class per file, but in some cases this encapsulates multiple structures/classes that are around one central idea (e.g. NSGeometry.swift).

Any further additions to the Foundation overlay ought to follow the pattern of an isolated file-scope and should keep the CMakeList.txt sorted.

### CODE REVIEW

**Comment by phausler** [comment]
@swift-ci Please test

**Comment by phausler** [comment]
@swift-ci Please test

**Comment by phausler** [comment]
This commit did not change, however the unit tests started now failing on macOS? I have verified that this is definitely a non-functional change; no code differs except for file locations.

**Comment by phausler** [comment]
@swift-ci Please clean test

**Comment by phausler** [comment]
@swift-ci Please test

**Comment by phausler** [comment]
Huzzah! a clean build; @jrose-apple I would like to get some reviewing eyes on this before merging if at all possible.

File: stdlib/public/SDK/Foundation/NSNumber.swift.gyb [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+// Numbers
+//===----------------------------------------------------------------------===//
+
+@_silgen_name("_swift_Foundation_TypePreservingNSNumberGetKind")
+internal func _swift_Foundation_TypePreservingNSNumberGetKind(
+  _ value: NSNumber
+) -> UInt8
+
+// This enum has a matching counterpart in TypePreservingNSNumber.mm. Please
+// update both copies when changing it.
```
Comment (jrose-apple) [inline stdlib/public/SDK/Foundation/NSNumber.swift.gyb]: There's a counterpart comment here as well, https://github.com/apple/swift/blob/master/stdlib/public/SDK/Foundation/TypePreservingNSNumber.mm#L26

**Review by jrose-apple — COMMENTED** [review]
I have opinions but nothing merge-blocking, and mostly I'm just happy to split up another monolith.
- NSSet and NSOrderedSet aren't really related, right?
- On the other hand, it might make sense to put the various collection features together, since they have similar code for each collection. That more closely matches the old organization, with all the bridging code, then all the hashing code, and so on. I certainly don't feel strongly about this, though.
- It seems weird to have, e.g., NSDate and Date code in different files. String and the collections are different because the value types and the reference types have independent existences, but all the Foundation-specific value types work closely enough with their reference types that I would expect them to appear together. That also helps with things like keeping the PlaygroundQuickLook behavior consistent.

**Comment by parkera** [comment]
@swift-ci please test and merge

**Comment by phausler** [comment]
Looks like I have been making the build bots work to hard and they decided to go take a coffee break.

```
15:41:57 Slave went offline during the build
15:41:57 ERROR: Connection was broken: java.io.IOException: Unexpected termination of the channel
```

**Comment by phausler** [comment]
@swift-ci please test and merge

**Comment by phausler** [comment]
Looks like cruft build data again :(

```
/home/buildnode/jenkins/workspace/swift-PR-Linux@2/branch-master/swiftpm/Sources/swift-build/main.swift:11:8: error: module file was created by an older version of the compiler; rebuild 'Commands' and try again: /home/buildnode/jenkins/workspace/swift-PR-Linux@2/branch-master/buildbot_linux/swiftpm-linux-x86_64/debug/Commands.swiftmodule
09:38:18 import Commands
```

@swift-ci please clean test and merge

**Pull request merged**